src/main/java/Board.java

```diff
@@ -9,9 +12,74 @@ public Grid getGrid() {
         return grid;
     }
 
-    public void addMarkToGrid(int position, String mark) {
+    public void takeSquare(int position, String playerMark) {
         int index = position - 1;
-        grid.setSquare(index, mark);
+        grid.setSquare(index, playerMark);
+    }
+
+    public boolean moveValid
+            (int position, String player1Mark, String player2Mark) {
+
+        return positionValid(position)
+                && squareAvailable(position, player1Mark, player2Mark);
+    }
+
+    public boolean winningLine(String playerMark) {
+        var winner = false;
+        var count = 0;
+        var size = 0;
+
+        for (int[] combination : grid.getWinningCombinations()) {
+            size = combination.length;
+
+            for (int index : combination) {
+                if (grid.getSquare(index) == playerMark) count += 1;
+                if (count == size) winner = true;
+            }
+            count = 0;
+        }
+        return winner;
+    }
+
+    public boolean winningLineExists
+            (String player1Mark, String player2Mark) {
+        return winningLine(player1Mark) || winningLine(player2Mark);
+    }
+
+    public boolean full(String player1Mark, String player2Mark) {
+        var availableSquares = availableSquares(player1Mark, player2Mark);
+        return availableSquares.size() == 0;
     }
 
+    public boolean complete(String player1Mark, String player2Mark) {
+        return full(player1Mark, player2Mark) ||
+                winningLineExists(player1Mark, player2Mark);
+    }
+
+    public List<String> availableSquares(String player1Mark, String player2Mark) {
+        List<String> availableSquares = new ArrayList<>();
```

Comment:
This code is all right (having the List and pushing into it inside of the for loop). If you want to check out another way to do it, google `filter` in Java. Here is a very similar example to the code you're working on: https://www.javacodegeeks.com/2018/07/filter-method-java-8.html

Reply:
Thanks! I'll look into this